tools/libeosshellcontent.py

```diff
@@ -0,0 +1,199 @@
+        metadata = self._get_app_metadata(app_id)
+
+        if not metadata:
+            app_id = app_id.rsplit('.desktop', 1)[0]
+            metadata = self._get_app_metadata(app_id)
+
+            if not metadata:
+                return False
+
+        app.set_name('C', metadata['title'])
+        app.set_comment('C', metadata['subtitle'])
+        app.set_description('C', metadata['description'])
+
+        categories = metadata.get('category')
+        for category in categories.split(';'):
+            app.add_category(category)
+
+        screenshot = AppStreamGlib.Screenshot()
+        for locale, images in metadata.get('screenshots', {}).items():
+            for image in images:
+                as_img = AppStreamGlib.Image()
+                as_img.set_url(self._get_screenshot_url(app_id, locale, image))
+                if locale != 'C':
+                    as_img.set_locale(locale)
+                screenshot.add_image(as_img)
+        if screenshot.get_images():
+            app.add_screenshot(screenshot)
+
+        self._translate_app(app)
+        return True
+
+    def _remove_unneeded_screenshots(self, xml_root):
+        images = xml_root.findall('./screenshots/screenshot/image')
+        has_shell_content_screenshots = False
+        for image in images:
+            if image.text.startswith(CMS_GS_BUCKET_URL):
+                has_shell_content_screenshots = True
+                break
+
+        # Do not remove screenshots unless we have some of ours specified
+        if not has_shell_content_screenshots:
+            return
+
+        # Remove all screenshots that had only images not starting with our URL
+        screenshots = xml_root.find('screenshots')
+        for screenshot in screenshots.getchildren():
+            for image in screenshot.findall('image'):
+                if not image.text.startswith(CMS_GS_BUCKET_URL):
+                    screenshot.remove(image)
+            if not screenshot.findall('image'):
+                screenshots.remove(screenshot)
+
+    def update_appdata_from_file(self, file_path, content_app_id=''):
+        app = load_as_app_from_appdata(file_path)
+        if not self.update_app(app, content_app_id):
+            return ''
+
+        store = AppStreamGlib.Store()
+        store.add_app(app)
+        xml_str = store.to_xml(AppStreamGlib.NodeToXmlFlags.NONE).str
+        root = ET.fromstring(xml_str)
+        if root.tag == 'components':
+            root = root.find('component')
+
+        # We do not want to display screenshots from 3rd parties when we have
+        # provided some from the Shell Content, so we need to remove them.
+        # The removal should be done in the app object but this cannot be
+        # accomplished until there is API to do it in libappstream-glib.
+        self._remove_unneeded_screenshots(root)
+
+        node = xml.dom.minidom.parseString(ET.tostring(root, encoding='utf-8'))
+        return node.toprettyxml(indent='  ')
```

Comment:
This was just for having pretty XML (indented). I searched how to do that with ElementTree but this version (as opposed to lxml's) doesn't have a direct way to prettify the output, so many people were suggesting just using the minidom for it. It's ugly code but we'll get pretty XML for a small price.